File: web/src/lib/bulkAvailability.ts
/**
 * Pure client-side preview for the "Set availability…" bulk dialog (plan §21
 * B2). Mirrors the *display-relevant* subset of the server's
 * `expandBulkSoloStatusSessions` (`firebase/functions/src/entries/bulkSoloStatus.ts`):
 * weekday match, date range clamped to today, and `kind !== 'noBridge'` —
 * enough to tell the member "this will touch N sessions, M of which are
 * already booked and will be left alone" before they confirm.
 *
 * Deliberately does *not* replicate the server's lock check or the 200-session
 * cap: the plan settles that the server enforces both and the preview can say
 * "about N" — reproducing session-cutoff timing client-side would just be
 * another place for the two to drift apart.
 */
import { todayNZ, type Entry, type IsoDate, type Session, type Weekday } from '@obc/shared';
import { isBookedEntry } from './overview';

export interface BulkAvailabilityFilter {
  weekdays: readonly Weekday[];
  fromDate?: IsoDate;
  toDate?: IsoDate;
}

export interface BulkAvailabilityPreview {
  /** Every session the filter matches (weekday + date range + bookable). */
  matched: number;
  /** Of `matched`, how many already have a booked entry and will be left untouched. */
  bookedSkipped: number;
  /** `matched - bookedSkipped` — sessions the action will actually change. */
  toUpdate: number;
}

/** The sessions `previewBulkAvailability` would count as "matched" — exposed for the dialog's date-list summaries. */
export function matchingSessions(sessions: readonly Session[], filter: BulkAvailabilityFilter, today: IsoDate = todayNZ()): Session[] {
  const weekdaySet = new Set<Weekday>(filter.weekdays);
  const from = filter.fromDate && filter.fromDate > today ? filter.fromDate : today;
  return sessions
    .filter((s) => weekdaySet.has(s.weekday))
    .filter((s) => s.kind !== 'noBridge')
    .filter((s) => s.date >= from)
    .filter((s) => !filter.toDate || s.date <= filter.toDate)
    .sort((a, b) => (a.date === b.date ? a.id.localeCompare(b.id) : a.date.localeCompare(b.date)));
}

export function previewBulkAvailability(
  sessions: readonly Session[],
  entries: readonly Entry[],
  filter: BulkAvailabilityFilter,
  today: IsoDate = todayNZ(),
): BulkAvailabilityPreview {
  const matched = matchingSessions(sessions, filter, today);
  const bookedSkipped = matched.filter((s) => {
    const entry = entries.find((e) => e.sessionId === s.id && e.status !== 'cancelled');
    return entry != null && isBookedEntry(entry);
  }).length;
  return { matched: matched.length, bookedSkipped, toUpdate: matched.length - bookedSkipped };
}

File: ios/OBCDanceCard/Features/Calendar/BulkAvailability.swift
//
//  BulkAvailability.swift
//  Pure client-side preview for the "Set availability…" bulk sheet — a 1:1
//  port of `web/src/lib/bulkAvailability.ts` (plan §21 B2). Mirrors the
//  display-relevant subset of the server's `expandBulkSoloStatusSessions`:
//  weekday match, date range clamped to today, and `kind != noBridge` —
//  enough to tell the member "this will touch N sessions, M of which are
//  already booked and will be left alone" before they confirm.
//
//  Deliberately does *not* replicate the server's session-lock check or its
//  200-session cap: the plan settles that the server enforces both and the
//  preview may say "about N".
//

import Foundation

struct BulkAvailabilityFilter: Hashable {
    var weekdays: Set<Weekday>
    var fromDate: String?
    var toDate: String?
}

struct BulkAvailabilityPreview: Equatable {
    /// Every session the filter matches (weekday + date range + bookable).
    var matched: Int
    /// Of `matched`, how many already have a booked entry and will be left untouched.
    var bookedSkipped: Int
    /// `matched - bookedSkipped` — sessions the action will actually change.
    var toUpdate: Int
}

enum BulkAvailability {
    /// The sessions the preview counts as "matched".
    static func matchingSessions(
        _ sessions: [Session],
        filter: BulkAvailabilityFilter,
        today: String = NZDate.today()
    ) -> [Session] {
        let from: String = {
            if let f = filter.fromDate, f > today { return f }
            return today
        }()
        return sessions
            .filter { filter.weekdays.contains($0.weekday) }
            .filter { $0.kind != .noBridge }
            .filter { $0.date >= from }
            .filter { s in filter.toDate.map { to in s.date <= to } ?? true }
            .sorted { a, b in a.date == b.date ? a.id < b.id : a.date < b.date }
    }

    static func preview(
        sessions: [Session],
        entries: [Entry],
        filter: BulkAvailabilityFilter,
        today: String = NZDate.today()
    ) -> BulkAvailabilityPreview {
        let matched = matchingSessions(sessions, filter: filter, today: today)
        let bookedSkipped = matched.filter { s in
            guard let e = entries.first(where: { $0.sessionId == s.id && $0.status != .cancelled }) else { return false }
            return Overview.isBookedEntry(e)
        }.count
        return BulkAvailabilityPreview(
            matched: matched.count,
            bookedSkipped: bookedSkipped,
            toUpdate: matched.count - bookedSkipped
        )
    }
}
